compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUBreakupMultiResultReductions.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright 2023 The IREE Authors
```

Comment:
I would prefer if we hold off on this. This is very adhoc and duplicating functionality from upstream (granted the usptream decompose linalg ops doesnt cover this case yet, but that needs to be fixed).

Reply:
Is the concern just the code splitting the linalg op? This is only few lines of code and I think what I need is different than what is upstream as ideally I want to control what gets broken down to only handle the case of reduction I need.
I can fix upstream if you think it is better but I'm not sure that's really what I want here.

Reply:
Concern is that this is duplicating functionality for a specific use case.... I'd happy to discuss why the upstream one is not what you want. I think when finished upstream, that is exactly what you are looking for. In any case, I'd rather hold off on this to first try and see if we can get away with not creating such a fused op in the first place. All of this would be not required then.

Reply:
I'm fine holding off if we want to generate different dispatch region. If this is the case it would be nice to add it to the roadmap to unlock the perf. Some of the changes would still be needed based on what the region would look like but it can go after.

Reply:
Let's figure out if the right thing is to finish the upstream one and go from there. It sounds like that may not be the case and resolving that will yield the path forward.

Reply:
yes based on offline discussions with Mahesh we agreed that the right next step is to change fusion to not generate such ops.

Reply:
> yes based on offline discussions with Mahesh we agreed that the right next step is to change fusion to not generate such ops.

Could you elaborate ?
Is the plan to not allow parallel + reduction dimensions to be collapsed into a reduction at all by IREE ?

To make it concrete, is IREE going to favor a softmax that resembles [this 6 ops form](https://github.com/iree-org/iree/blob/main/tests/transform_dialect/cuda/softmax.mlir#L85)
instead of a softmax that resembles [this 5 ops form](https://github.com/iree-org/iree/blob/main/tests/transform_dialect/cuda/softmax_v2.mlir#L63) (which I think is close to what current happens) ?

Or am I misinterpreting things ?

Reply:
> > yes based on offline discussions with Mahesh we agreed that the right next step is to change fusion to not generate such ops.
> 
> Could you elaborate ? Is the plan to not allow parallel + reduction dimensions to be collapsed into a reduction at all by IREE ?
> 

Correct.

> To make it concrete, is IREE going to favor a softmax that resembles [this 6 ops form](https://github.com/iree-org/iree/blob/main/tests/transform_dialect/cuda/softmax.mlir#L85) instead of a softmax that resembles [this 5 ops form](https://github.com/iree-org/iree/blob/main/tests/transform_dialect/cuda/softmax_v2.mlir#L63) (which I think is close to what currently happens) ?
> 

pretty much although I think we may still want simple merge for parallel ops (could be down to 3 ops in this case)

To give more details we are going to explore two potential directions. First Mahesh is going to try to change the default fusion logic to support this dag kind of fusion. If this turns out to be too complicated we would add a pre-fusion pass to catch special cases we want to fuse. Based on discussions with Mahesh, we will want this mechanism at some point anyway to be able to catch for complex and specific fusions.

Reply:
Great, thanks for confirming.

Avoiding fusing parallel + red into a multi-result op is going to make things simpler to reason about so big +1 from me, I strongly prefer the 6 ops form to the 5 ops form.

Reply:
https://github.com/iree-org/iree/pull/11740 seems to work and do what is requested here. Hopefully that puts an end to the long running saga of softmax.
Two things seem to come out of this though
1) It seems backwards that we have the front end break up softmax and then we are putting all this effort to lift the DAG back into softmax. This seems to be another example of a op that decomposes into a list of ops. To repeat what I have been saying from the start, we should just add a softmax op that implements the TilingInterface, get the front end to lower to this op and then break the op up when needed. We have already a precedence of this kind of thing working based on Winograd. That seems a better approach (and then we dont need to worry about the 4-op version, 5-op version or 6-op version). There are other ops in the pipeline (like flash-attention) that seem to merit a similar approach
2) Controversial statement, but I think the softmax handling has been special cased to the specific sequence of ops expected. On the CPU side for example fused parallel + red multi-result op suffers from the same issue as the non fused version (https://github.com/iree-org/iree/issues/11412). More the reason to have a dedicated op for softmax instead of special-casing based on a DAG of operations.

Reply:
The need for a sequence-like op and the associated named <-> generic mechanism has been identified for a very long time, if this becomes critical enough that someone wants to start building generic reusable infra for this, I'd welcome it warmly :)

Reply:
@nicolasvasilache lets chat about what you have in mind....

Reply:
Thanks Mahesh! I cherry-picked your change and it works great. 

> 2\. Controversial statement, but I think the softmax handling has been special cased to the specific sequence of ops expected.

This is a fine line but avoiding storing in temporary buffer is in general a good thing and this is what is done here. Especially since we know reductions are memory bound it will be better to re-materialize. I don't think it is that specific of a case.

> 2\. On the CPU side for example fused parallel + red multi-result op suffers from the same issue as the non fused version ([Avoid creating allocation during bufferization #11412](https://github.com/iree-org/iree/issues/11412)). 

It is pretty easy to build a similar example where we cannot store in the buffer (for instance if there a type change in the middle preventing to reuse the output buffer). So avoiding having those temp allocation feels like a general problem.

Reply:
> It is pretty easy to build a similar example where we cannot store in the buffer (for instance if there a type change in the middle preventing to reuse the output buffer). So avoiding having those temp allocation feels like a general problem.

Not saying it is not a general problem. I am saying that the problem exists whether we create a single op with two results or leave as two separate ops. Its a bufferization issue. The rematerialization being done is what I think is one-off (cause in general when to rematerialize is hard to model)